plenum/server/replica.py

```diff
@@ -774,9 +778,19 @@ def processReqDuringBatch(
             self.logger.warning('{} encountered exception {} while processing {}, '
                                 'will reject'.format(self, ex, req))
             rejects.append((req.key, Reject(req.identifier, req.reqId, ex)))
-            inValidReqs.append(req)
+            discarded_mask.append(True)
         else:
-            validReqs.append(req)
+            discarded_mask.append(False)
+        finally:
+            reqs.append(req)
+
+    def _pack_discarded_mask(self, value):
```

Comment:
pack_discarded_mask, unpack_discarded_mask, drop_discarded_mask, apply_bitmask_to_list, get_valid_reqs, get_invalid_reqs, get_invalid_count needs to be moved to a separate utility class (not in Replica)